core/src/main/java/org/apache/iceberg/avro/GenericAvroReader.java

```diff
@@ -155,6 +162,41 @@ public ValueReader<?> record(Type partner, Schema record, List<ValueReader<?>> f
       return recordReader(readPlan, avroSchemas.get(partner), record.getFullName());
     }
 
+    Object toGenericRecord(Type type, Object data) {
+      // Recursively convert data to GenericRecord if type is a StructType.
+      // TODO: Rewrite this as a visitor.
```

Comment:
Converting this into a `TypeUtil.SchemaVisitor` would make sense

Reply:
While I initially intended to implement it as a visitor, it turns out none of the existing visitor patterns (`SchemaVisitor` or `SchemaWithPartnerVisitor`) is a good fit for this use case since here we are co-traversing a schema and a data object (not just a schema or two schemas). I will clean up this code instead of using a visitor.